### PR TITLE
Fix `searchsorted` with non-contiguous out

### LINKED ISSUE #60492: searchsorted produces incorrect output if out is given explicitly and not contiguous
## 🐛 Bug

`torch.searchsorted` has an `out` parameter but it assumes it is contiguous. Otherwise an incorrect result is produced without any warnings.

## To Reproduce

```python
import torch
device = torch.device("cpu")
# device = torch.device("cuda")

x = torch.arange(10, device=device)
ind = torch.arange(3, 8, device=device)
tmp = torch.zeros(5, 2, device=device, dtype=ind.dtype)
out = tmp[:, 0]

torch.searchsorted(x, ind, out=out)
print(out)
print(tmp)
```

CPU output:
```
tensor([3, 5, 7, 0, 0])
tensor([[3, 4],
        [5, 6],
        [7, 0],
        [0, 0],
        [0, 0]])
```

CUDA output:
```
tensor([3, 5, 7, 0, 0], device='cuda:0')
tensor([[3, 4],
        [5, 6],
        [7, 0],
        [0, 0],
        [0, 0]], device='cuda:0')
```

## Expected behavior

It can print a warning, raise an exception, convert `out` to a contiguous tensor or consider its strides. Not really sure. This behavior can be noted in docs.

## Environment

python 3.8.8
pytorch 1.8.1
cudatoolkit 11.1.1
via conda

cc @heitorschueroff

### PR DESCRIPTION
**Pull request by emcastillo**
Fixes  #60492

TODO: add test

### CODE REVIEW

File: aten/src/ATen/native/Bucketization.cpp [hunk 1]
```diff
@@ -105,7 +112,10 @@ Tensor& searchsorted_out_cpu(const Tensor& sorted_sequence, const Tensor& self,
   searchsorted_maybe_trim_input_tensors(trimmed_input, trimmed_boundaries, self, sorted_sequence);
   const Tensor& final_input = trimmed_input.defined() ? trimmed_input : self;
   const Tensor& final_boundaries = trimmed_boundaries.defined() ? trimmed_boundaries : sorted_sequence;
-  dispatch(result, final_input, final_boundaries, out_int32, right);
+  dispatch(out, final_input, final_boundaries, out_int32, right);
+  if (!out.is_contiguous()) {
```
Comment (heitorschueroff) [inline aten/src/ATen/native/Bucketization.cpp:116]: out -> result

File: aten/src/ATen/native/Bucketization.cpp [hunk 2]
```diff
@@ -105,7 +112,10 @@ Tensor& searchsorted_out_cpu(const Tensor& sorted_sequence, const Tensor& self,
...
+    result.copy_(out);
+  }
   return result;
```
Comment (heitorschueroff) [inline aten/src/ATen/native/Bucketization.cpp:119]: Can you combine lines 104-107 above with these ones so there's a single return statement. The two different dispatch calls can be in an if-else statement.

File: aten/src/ATen/native/Bucketization.cpp [hunk 3]
```diff
@@ -89,14 +89,21 @@ void dispatch(Tensor& result, const Tensor& input, const Tensor& boundaries, boo
 
 Tensor& searchsorted_out_cpu(const Tensor& sorted_sequence, const Tensor& self, bool out_int32, bool right, Tensor& result) {
   searchsorted_pre_check(sorted_sequence, self, result, out_int32);
+  Tensor out = result;
   if (result.numel() == 0) {
     result.resize_(self.sizes());
   }
+  if (!result.is_contiguous()) {
+    out = result.contiguous();
+  }
```
Comment (heitorschueroff) [inline aten/src/ATen/native/Bucketization.cpp:98]: Can you move lines 92 and 96-98 after line 101 which is when we do need to create the out tensor?

File: aten/src/ATen/native/Bucketization.cpp [hunk 4]
```diff
@@ -89,14 +89,21 @@ void dispatch(Tensor& result, const Tensor& input, const Tensor& boundaries, boo
 
 Tensor& searchsorted_out_cpu(const Tensor& sorted_sequence, const Tensor& self, bool out_int32, bool right, Tensor& result) {
   searchsorted_pre_check(sorted_sequence, self, result, out_int32);
+  Tensor out = result;
   if (result.numel() == 0) {
     result.resize_(self.sizes());
```
Comment (heitorschueroff) [inline aten/src/ATen/native/Bucketization.cpp:94]: We should use `at::native::resize_output` for resizing output tensors passed with `out=` parameter. It already checks that either the `out` tensor is empty or it's size is correct with the desired size.

File: aten/src/ATen/native/cuda/Bucketization.cu [hunk 5]
```diff
@@ -106,14 +106,21 @@ void dispatch(Tensor& result, const Tensor& input, const Tensor& boundaries, boo
 
 Tensor& searchsorted_out_cuda(const Tensor& sorted_sequence, const Tensor& self, bool out_int32, bool right, Tensor& result) {
   searchsorted_pre_check(sorted_sequence, self, result, out_int32);
+  Tensor out = result;
   if (result.numel() == 0) {
     result.resize_(self.sizes());
```
Comment (heitorschueroff) [inline aten/src/ATen/native/cuda/Bucketization.cu:111]: Same as above, use at::native::resize_output

File: aten/src/ATen/native/cuda/Bucketization.cu [hunk 6]
```diff
@@ -106,14 +106,21 @@ void dispatch(Tensor& result, const Tensor& input, const Tensor& boundaries, boo
 
 Tensor& searchsorted_out_cuda(const Tensor& sorted_sequence, const Tensor& self, bool out_int32, bool right, Tensor& result) {
   searchsorted_pre_check(sorted_sequence, self, result, out_int32);
+  Tensor out = result;
```
Comment (heitorschueroff) [inline aten/src/ATen/native/cuda/Bucketization.cu:109]: nit: we usually want to declare variables closest to when we need them, could you move this before line 116

File: aten/src/ATen/native/cuda/Bucketization.cu [hunk 7]
```diff
@@ -122,7 +129,10 @@ Tensor& searchsorted_out_cuda(const Tensor& sorted_sequence, const Tensor& self,
   searchsorted_maybe_trim_input_tensors(trimmed_input, trimmed_boundaries, self, sorted_sequence);
   const Tensor& final_input = trimmed_input.defined() ? trimmed_input : self;
   const Tensor& final_boundaries = trimmed_boundaries.defined() ? trimmed_boundaries : sorted_sequence;
-  dispatch(result, final_input, final_boundaries, out_int32, right);
+  dispatch(out, final_input, final_boundaries, out_int32, right);
+  if (!result.is_contiguous()) {
+    result.copy_(out);
+  }
```
Comment (heitorschueroff) [inline aten/src/ATen/native/cuda/Bucketization.cu:135]: Same as above, could you combine the two dispatch cases in an if-else and have a single is_contiguous check?

**Review by heitorschueroff — CHANGES_REQUESTED** [review]
@emcastillo Thank you for fixing this bug. You have a small bug where you're checking for contiguity in out instead of result. And although not part of your changes, would you replace the resize call with a call to `at::native::resize_output` which is the preferred way to resize out tensors. Besides that, mostly stylistic changes.

Also, could you create an OpInfo test for `searchsorted`. The OpInfo should automatically test for the`out=` parameter and should fail for the current implementation which does not work for discontiguous out tensors.

Feel free to ping me if you have any questions and I look forward to the next update. Thanks!

**Comment by emcastillo** [comment]
Hi!
Thanks for the review!
I am currently looking info regarding the `OpInfo` tests but I can only find TODOs in the tests/ folder 😅

**Comment by heitorschueroff** [comment]
> I see the implementation in `toirch/testing/_internal/common_methods_invocations.py` but should the `searchsorted` test go there?

Good question! All OpInfo tests currently go into the list here https://github.com/pytorch/pytorch/blob/552eab7935c44d5f04b727fc4914559d6029ea64/torch/testing/_internal/common_methods_invocations.py#L4776

You can read through a few examples as well as the comment on the OpInfo class. Let me know if you have any questions regarding the tests, I am happy to answer them. Once you'd like me to take another look at the PR please request my review again. Thanks!

**Pull request closed without merging**